components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -89,6 +89,72 @@ class StyxObjectStore<T> : ObjectStore<T> {
         return Optional.ofNullable(current[key])
     }
 
+    /**
+     * Insert an object in the database that is a result of a function call.
+     *
+     * This method can be used to insert a new object, or to conditionally modify
+     * a previously stored object.  This method takes two arguments: `key` and
+     * `computation`.
+     *
+     * The `key` refers to the object name of interest. The `computation` provides
+     * a new or modified object to be stored under `key`.
+     *
+     * `computation` is called with an existing object as its only argument, or
+     * `null` if no objects are stored against `key`. An object returned from
+     * `computation` is stored in the database.
+     *
+     * The operation is considered a no-op when `computation` returns back the
+     * already existing object instance. In this case the existing object remains
+     * in the database, and watchers are not notified.
+     *
+     * When `computation` returns a different instance, the new object replaces
+     * the existing version and the watchers will be notified. This is considered
+     * a modification.
+     *
+     * Note that the `computation` must always return a value. It cannot return
+     * `null`, and therefore cannot be used to remove existing objects.
+     *
+     * @property key object name
+     * @property computation a function that produces the new value
+     * @return the previous value
+     */
+    fun compute(key: String, computation: (T?) -> T): Optional<T> {
+        require(key.isNotEmpty())
+
+        var current = objects.get()
+        var result = computation(current.get(key))
+
+        var new = if (result != current.get(key)){
+            // Consumer REPLACES an existing value or ADDS a new value
+            current.plus(key, result)
+        } else {
+            // Consumer KEEPS the existing value
+            current
+        }
+
+        while(!objects.compareAndSet(current, new)) {
```

Comment:
The `current`, `result` and `new` would have to be declared outside of do/while scope, defeating the purpose.

At the moment we merely initialise them on declaration. However the initialisation for `new` is rather complicated with 3 lines of code. So, perhaps that would be a small win.

Reply:
The code would look like this. WDYT?

```java
    fun compute(key: String, computation: (T?) -> T): Optional<T> {
        require(key.isNotEmpty()) { "ObjectStore compute: empty keys are not allowed." }

            var current: PMap<String, T>
            var result: T
            var new: PMap<String, T>

            do {
                current = objects.get()
                result = computation(current.get(key))

                new = if (result != current.get(key)){
                    // Consumer REPLACES an existing value or ADDS a new value
                    current.plus(key, result)
                } else {
                    // Consumer KEEPS the existing value
                    current
                }
            } while(!objects.compareAndSet(current, new))

            if (current != new) {
                // Notify only if content changed:
                queue {
                    notifyWatchers(new)
                }
            }

            return Optional.ofNullable(current[key])
    }
```

Reply:
MMm, I think it looks a bit neater but you are right it doesn't make much of a difference. 

What do you think of abstracting the while computation -it could receive a lambda I guess - in its own method since all the other methods are doing the  !objects.compareAndSet(current, new) ? Not important anyway.

Reply:
There are many other more pressing issues. Therefore I'll leave this like it is :-)